src/v/storage/log_manager.cc

```diff
@@ -88,18 +88,44 @@ ss::future<> log_manager::stop() {
     return _open_gate.close()
       .then([this] {
           return ss::parallel_for_each(_logs, [](logs_type::value_type& entry) {
-              return entry.second.handle.close();
+              return entry.second->handle.close();
           });
       })
       .then([this] { return _batch_cache.stop(); });
 }
 
-static inline logs_type::iterator find_next_non_compacted_log(logs_type& logs) {
+/**
+ * `housekeeping_scan` scans over every current log in a single pass.
+ */
+ss::future<>
+log_manager::housekeeping_scan(model::timestamp collection_threshold) {
     using bflags = log_housekeeping_meta::bitflags;
-    return std::find_if(
-      logs.begin(), logs.end(), [](const logs_type::value_type& l) {
-          return bflags::none == (l.second.flags & bflags::compacted);
-      });
+
+    if (_logs_list.empty()) {
+        co_return;
+    }
+
+    while ((_logs_list.front().flags & bflags::compacted) == bflags::none) {
```

Comment:
is there a possibility of a scenario occurring where the first item doesn't need to be compacted, but the second element does and this loop won't run and find the second element? this seems very error prone in order to maintain this property. perhaps we should still scan the whole list? the alternative is to maintain data structures that serve as indicies over exactly the type of query we want to execute (e.g. GC/retention time is due, etc...)

Reply:
The bflags::compacted might be a misnomer in this case. It doesn't really indicate if the log has been compacted or not. I believe that the logic to determine if the log should be compacted is in the log's compact function. The log_manager is just responsible for ensuring that function is called periodically on every log.

To that end the compacted flag is cleared on all logs before the housekeeping_scan starts then it is set one at a time as compacted is called on the log. So then the housekeeping_scan reaches a log where the flag has already been set then we know we've called compact on every log.

So in essence we are scanning the entire list each time. But rather than holding an iterator we are marking a flag on each element we read.

Reply:
Oh of course. I see at the end of housekeeping all the flags get cleared again

```
      .finally([this] {
          for (auto& h : _logs) {
              h.second.flags &= ~bflags::compacted;
          for (auto& h : _logs_list) {
              h.flags &= ~bflags::compacted;
          }
```

thanks for reminding me!

Reply:
ill resolve this it isn't an issue at all.